src/storage/leveldb/address_index.ts

```diff
@@ -0,0 +1,184 @@
+
+  constructor(dbpath: string) {
+    this.dbpath = path.join(dbpath, 'addresses');
+    // Open addresses
+    const db = new Level(this.dbpath);
+    this.addressesDB = db.sublevel<string, IAddressInfo>(ADDRESS_PREFIX, {valueEncoding: 'json'});
+    this.addressesIndexDB = db.sublevel(INDEX_PREFIX);
+    this.addressesMetaDB = db.sublevel<string, IAddressMetadataAsRecord>(ADDRESS_META_PREFIX, {valueEncoding: 'json'});
+  }
+
+  async close(): Promise<void> {
+    await this.addressesDB.db.close();
+  }
+
+  async checkVersion(): Promise<void> {
+    const db = this.addressesDB.db;
+    try {
+      const dbVersion = await db.get('version');
+      if (this.indexVersion !== dbVersion) {
+        throw new Error(`Database version mismatch for ${this.constructor.name}: database version (${dbVersion}) expected version (${this.indexVersion})`);
+      }
+    } catch (err: unknown) {
+      if (errorCodeOrNull(err) === KEY_NOT_FOUND_CODE) {
+        // This is a new db, add version and return
+        await db.put('version', this.indexVersion);
+        return;
+      }
+
+      throw err;
+    }
+  }
+
+  async validate(): Promise<AddressIndexValidateResponse> {
+    // Clear metadata since we cannot guarantee the validity
+    await this.addressesMetaDB.clear();
+
+    const ret: AddressIndexValidateResponse = {firstIndex: Infinity, lastIndex: -1};
+    // Iterate on all addresses and check that we have a corresponding index entry
+    for await (const [key, value] of this.addressesDB.iterator()) {
+      if (key !== value.base58) {
+        throw new Error('Inconsistent database');
+      }
+      if (value.bip32AddressIndex < ret.firstIndex) {
+        ret.firstIndex = value.bip32AddressIndex;
+      }
+      if (value.bip32AddressIndex > ret.lastIndex) {
+        ret.lastIndex = value.bip32AddressIndex;
+      }
+
+      // check that we have an index
+      try {
+        const addressFromIndex = await this.addressesIndexDB.get(_index_key(value.bip32AddressIndex));
+        if (value.base58 !== addressFromIndex) {
+          throw new Error('Inconsistent database');
+        }
+      } catch(err: unknown) {
+        if (errorCodeOrNull(err) === KEY_NOT_FOUND_CODE) {
+          // Create if index is missing
+          await this.addressesIndexDB.put(_index_key(value.bip32AddressIndex), value.base58);
+          continue;
+        }
+        throw err;
+      }
+    }
+    return ret;
+  }
+
+  async getAddressInfo(base58: string): Promise<IAddressInfo | null> {
+    try {
+      return await this.addressesDB.get(base58);
+    } catch(err: unknown) {
+      if (errorCodeOrNull(err) === KEY_NOT_FOUND_CODE) {
+        return null;
+      }
+      throw err;
+    }
+  }
+
+  async addressExists(base58: string): Promise<boolean> {
+    return (await this.getAddressInfo(base58)) !== null;
+  }
+
+  async * addressIter(): AsyncGenerator<IAddressInfo> {
+    for await (const address of this.addressesIndexDB.values()) {
```

Comment:
Because of how leveldb handles iteration, all entries are sorted by the key value.
Which means that when I iterate on the `addressIndexDB` the order of the returned addresses will be the same as the order of the keys of `addressIndexDB` and since we use the uint32 (in hex format) the addresses will returned in order of the bip32 address index